### PR TITLE
DS-3086 OAI Harvester is broken

### PR DESCRIPTION
**Pull request by DylanMeeus**
https://jira.duraspace.org/browse/DS-3086

Fixing changes that broke the OAI-harvester

### CODE REVIEW

**Comment by helix84** [comment]
I tested the PR. I can confirm the problem. This fix makes things better in that metadata formats are populated again, test settings reports success and harvest can be run. The harvest seems to run successfully (I tested with the sample collection from demo.dspace.org), no errors in the log. However, no items appear in the collection.

When I press Reset and reimport, I get the following error:
ERROR: update or delete on table "item" violates foreign key constraint "harvested_item_item_id_fkey" on table "harvested_item" Detail: Key (uuid)=(18b6cdbb-a777-4668-a4d3-321824a0dea4) is still referenced from table "harvested_item".

**Comment by helix84** [comment]
This error appeared in the log after I tried the failed re-import and after I pressed Return to go back from Edit collection to the collection: http://pastebin.com/E8ccZmx4

**Comment by DylanMeeus** [comment]
@helix84 I will check it out, thanks!

**Comment by helix84** [comment]
I think I found the culprit - I had to run index-discovery and the items appear now.
Not sure if you can reprocude this - but before the reindex refreshing the empty collection threw the "transaction aborted" stacktrace again.

**Comment by DylanMeeus** [comment]
@helix84 I can confirm that I could recreate the problems that you have mentioned.

**Comment by tdonohue** [comment]
@DylanMeeus : this PR looks to need a rebase, as it has conflicts with the latest master

**Comment by DylanMeeus** [comment]
@tdonohue Thanks, I will do so

**Comment by DylanMeeus** [comment]
Renamed this PR because it's more than just NullPointerExceptions. 
With the current fixes, the import and reimport already work as expected but noticed some other minor issues. Will update the branch soon.

**Comment by tomdesair** [comment]
This PR can be considered as complete now. An explanation of all the changes can be found here: https://jira.duraspace.org/browse/DS-3086?focusedCommentId=49448&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-49448

**Comment by tdonohue** [comment]
This looks to have merge conflicts with latest master. Flagging as such.

**Comment by DylanMeeus** [comment]
Rebased on current master.
All checks passed an no more merge conflicts exist

**Comment by mwoodiupui** [comment]
Do you have a moment to de-conflict this one more time, please?

**Comment by DylanMeeus** [comment]
Will do so, but it will be after the weekend (Tuesday, monday is a holiday here)

File: dspace-api/src/main/java/org/dspace/content/crosswalk/CrosswalkMetadataValidator.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            if (mdSchema == null) {
+                // add a new schema, giving it a namespace of "unknown". Possibly a very bad idea.
+                if (forceCreate && schemaChoice.equals("add")) {
+                    try {
+                        mdSchema = metadataSchemaService.create(context, schema, String.valueOf(new Date().getTime()));
+                        mdSchema.setNamespace("unknown" + mdSchema.getID());
+                        metadataSchemaService.update(context, mdSchema);
+                    } catch (NonUniqueMetadataException e) {
+                        // This case should not be possible
+                        e.printStackTrace();
```
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/content/crosswalk/CrosswalkMetadataValidator.java]: (1) User is not told why he is seeing a stack trace.  (2) Error output should go to System.err:  e.printStackTrace(System.err);

File: dspace-api/src/main/java/org/dspace/content/crosswalk/CrosswalkMetadataValidator.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            if (mdSchema != null) {
+                // Verify that the element exists; this part is reachable only if the metadata schema is valid
+                mdField = metadataFieldService.findByElement(context, mdSchema, element, qualifier);
+                if (mdField == null) {
+                    if (forceCreate && fieldChoice.equals("add")) {
+                        try {
+                            metadataFieldService.create(context, mdSchema, element, qualifier, null);
+                        } catch (NonUniqueMetadataException e) {
+                            // This case should also not be possible
+                            e.printStackTrace();
```
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/content/crosswalk/CrosswalkMetadataValidator.java]: Another mystery stack trace.

**Comment by mwoodiupui** [comment]
+1 with an asterisk.  I've read it all, and I think I understand it.  There is nothing obviously wrong, and there is so much change that we will need to have it beaten up with a real-world load to discover any remaining issues.

This apparently good work is a bug fix PR and an improvement PR rolled into one.  I read it straight through, but it might be easier to review commit-by-commit.  The actual bug fix is quite compact.

**Comment by hardyoyo** [comment]
I've made an attempt to test this code, by checking out the PR branch, and rebasing it on a current copy of master, and resolving the pretty basic merge conflict. The code compiles cleanly and deploys. However, in testing a harvest using the OAI client, I still encountered an NPE. https://gist.github.com/hardyoyo/4cc03e06b84c5e3bdacc6e451480eee2

**Comment by tdonohue** [comment]
Hi @DylanMeeus : Looks like this has a merge conflict. Any chance you could rebase it?
We seem to have other PRs which are related, and it's hard to tell if this one will resolve the other issues..see #1385 and #1381

**Comment by DylanMeeus** [comment]
@tdonohue rebased it. Haven't yet checked out that NPE @hardyoyo

**Comment by terrywbrady** [comment]
I configured a harvest from demo.dspace.org, and I see the following error on Import:

Java stacktrace: 
java.lang.NullPointerException
	at org.dspace.core.AbstractHibernateDAO.getHibernateSession(AbstractHibernateDAO.java:48)
	at org.dspace.core.AbstractHibernateDAO.findByID(AbstractHibernateDAO.java:75)
	at org.dspace.content.CollectionServiceImpl.find(CollectionServiceImpl.java:234)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)...

**Comment by pnbecker** [comment]
@DylanMeeus I'm sorry to say so, but we have a new merge conflict here. Beside @terrywbrady comment it needs another rebase.

**Comment by terrywbrady** [comment]
+1 based on testing.  I have successfully harvested TO and harvested FROM collections using this code.

**Comment by pnbecker** [comment]
This is +2 and was tested just some days ago by @terrywbrady. I'll merge this after short discussion with @mwoodiupui and @tdonohue in #dspace.

**Pull request merged**